ddtrace/internal/module.py

```diff
@@ -0,0 +1,427 @@
+from collections import defaultdict
+from os.path import abspath
+from os.path import expanduser
+from os.path import isdir
+from os.path import isfile
+from os.path import join
+import sys
+from types import ModuleType
+from typing import Any
+from typing import Callable
+from typing import DefaultDict
+from typing import Iterator
+from typing import List
+from typing import Optional
+from typing import Set
+from typing import TYPE_CHECKING
+from typing import Tuple
+from typing import Union
+
+from bytecode import Bytecode
+from bytecode import Instr
+
+from ddtrace.internal.compat import PY2
+from ddtrace.internal.logger import get_logger
+
+
+if TYPE_CHECKING and not PY2:
+    from importlib.abc import Loader
+
+log = get_logger(__name__)
+
+HookType = Callable[[ModuleType, Any], None]
+
+
+def origin(module):
+    # type: (ModuleType) -> str
+    """Get the origin of the module."""
+    try:
+        orig = abspath(module.__file__)  # type: ignore[type-var]
+    except (AttributeError, TypeError):
+        # Module is porbably only partially initialised, so we look at its
+        # spec instead
+        try:
+            orig = abspath(module.__spec__.origin)  # type: ignore
+        except (AttributeError, ValueError, TypeError):
+            orig = None
+
+    if orig is not None and isfile(orig):
+        if orig.endswith(".pyc"):
+            orig = orig[:-1]
+        return orig
+
+    return "<unknown origin>"
+
+
+def _resolve(path):
+    # type: (str) -> Optional[str]
+    """Resolve a (relative) path with respect to sys.path."""
+    for base in sys.path:
+        if isdir(base):
+            resolved_path = abspath(join(base, expanduser(path)))
+            if isfile(resolved_path):
+                return resolved_path
+    return None
+
+
+# Borrowed from the wrapt module
+# https://github.com/GrahamDumpleton/wrapt/blob/0baff1b6ac8d64ffb08ea2d1610c7ed90115b9d5/src/wrapt/importer.py
+class _ImportHookChainedLoader:
+    def __init__(self, loader, callback):
+        # type: (Loader, Callable[[ModuleType], None]) -> None
+        self.loader = loader
+        self.callback = callback
+
+    def load_module(self, fullname):
+        # type: (str) -> ModuleType
+        module = self.loader.load_module(fullname)
+        self.callback(module)
+
+        return module
+
+    def get_code(self, mod_name):
+        return self.loader.get_code(mod_name)
+
+
+class ModuleWatchdog(dict):
+    """Module watchdog.
+
+    Replace the standard ``sys.modules`` dictionary to detect when modules are
+    loaded/unloaded. This is also responsible for triggering any registered
+    import hooks.
+    """
+
+    _run_code = None
+
+    def __init__(self):
+        # type: () -> None
+        self._hook_map = defaultdict(list)  # type: DefaultDict[str, List[Tuple[HookType, Any]]]
+        self._origin_map = {origin(module): module for module in sys.modules.values()}
+        self._modules = sys.modules
```

Comment:
```suggestion
        self._modules = sys.modules  # type: Union[dict, ModuleWatchdog]
```

maybe to help readability